discourse: 2.7.9 -> 2.8.0.beta9

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
- Update Discourse to the latest beta release, as discussed in #146308
- Add a `libv8` output to `nodejs` as an alternative to `v8`, which is a lot of work to update
- Update Discourse plugins to the latest releases and make it possible to update `discourse-prometheus`
- Replace `up-plugins.sh` with a `README` pointing to the plugin packaging docs
- Improve version handling in the `update.py` script
- Implement Discourse's [plugin pinning algorithm](https://meta.discourse.org/t/pinning-plugin-and-theme-versions-for-older-discourse-installs/156971) in the `update.py` script

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

talyz committed:
nodejs: Provide a static v8 library output

Since building nodejs also builds v8, one way to get a static v8
library is to manually assemble it from the leftover object
files. This seems like an easier way to get an up-to-date v8 library
than trying to keep the v8 package updated.

talyz committed:
discourse: 2.7.9 -> 2.8.0.beta9

Update to the latest beta, since upstream advocates for it. See
https://github.com/NixOS/nixpkgs/issues/146308 for more info.

talyz committed:
discourse.plugins: Update all plugins to their latest versions

Also, add support for updating plugins which keep gem versions in
files at the root of the repo (discourse-prometheus) and replace the
`up-plugin.sh` script with a README file pointing to the plugin
packaging documentation.

ryantm reviewed (approved):
Code changes look good, as far as I can assess. I don't know much about the v8 changes.

I've upgraded a few of my deployments with it without trouble. Tested the following plugins work: LDAP, data-explorer, solved, checklist, calendar.

SuperSandro2000 commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -824,8 +827,14 @@ in
       upstreams.discourse.servers."unix:/run/discourse/sockets/unicorn.sock" = {};
 
       appendHttpConfig = ''
+        # Additional MIME types that nginx should handle
```
Our nginx uses mailcap mime.types which already includes those.

talyz replied:
Ah, indeed we do. Removed. :)

SuperSandro2000 commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -130,6 +134,35 @@ let
+      # assemble a static v8 library and put it in the 'libv8' output
+      mkdir -p $libv8/lib
+      pushd out/Release/obj.target
+      find . -path "./torque_*/**/*.o" -or -path "./v8*/**/*.o" | sort | uniq >files
```
```suggestion
      find . -path "./torque_*/**/*.o" -or -path "./v8*/**/*.o" | sort -u >files
```
sort can do that and maybe it is even faster for a lot of files.

talyz replied:
Yeah, good point. It's probably a bit faster.

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/plugins/discourse-prometheus/prometheus_exporter_version:
```diff
@@ -0,0 +1 @@
+0.5.0
```
It is not very obvious why we need this file. Could it be also generated on the fly?

talyz replied:
It's imported from the plugin's repo by the `update.py` script and read by the `Gemfile`.

talyz committed:
discourse: update.py: Improve version handling, use pinned plugins

Add a DiscourseVersion class which handles Discourse's version
numbering properly when sorting - beta versions are sorted lower than
their respective release versions. It can also return both its version
number and equivalent git tag, removing the need for `rev2version` and
manually adding `v` to the front.

Using DiscourseVersion instead of LooseVersion, we can list all
current version number tags from the `discourse` repo and sort them
correctly, giving us the latest one, regardless of type; i.e. we don't
have to filter for only release versions or beta versions anymore.

This also implements the plugin pinning algorithm laid out here:
https://meta.discourse.org/t/pinning-plugin-and-theme-versions-for-older-discourse-installs/156971
to make sure we don't upgrade plugins further than what's compatible
with our currently packaged Discourse version. While it likely won't
matter much most of the time if we continue packaging the beta
versions, it could be helpful if we decide to go back to packaging
release versions or if we run into issues with future upgrades. In
that case, the plugins could still be updated safely even though we're
not on the latest version of Discourse.

talyz commented:
I've updated to `2.8.0.beta9` and improved the version handling in the `update.py` script. It should now look for the latest release among _all_ available version numbered releases. It also now implements Discourse's [plugin pinning algorithm](https://meta.discourse.org/t/pinning-plugin-and-theme-versions-for-older-discourse-installs/156971) to make sure we only update plugins to versions which are compatible with our packaged version of Discourse. While this likely won't matter much most of the time if we continue packaging the beta versions, it could be helpful if we decide to go back to packaging
release versions or if we run into issues with future upgrades. In that case, the plugins could still be updated safely even though we're not on the latest version of Discourse.

ryantm commented:
Cool, @talyz. I deployed your latest thing to my deployments and it seems to be working fine.

talyz committed:
discourse: Don't patch the public path

Instead of patching the path to /public in Discourse's sources, make
the nginx configuration refer to the symlink in the discourse
package which points to the real path.

When there is a mismatch between the path nginx serves and the path
Discourse thinks it serves, we can run into issues like files not
being served - at least when sendfile requests from the ruby app are
processed by nginx. The issue I ran into most recently is that backup
downloads don't work.

Since Discourse refers to the public directory relative to the Rails
root in many places, it's much easier to just sync this path to the
nginx configuration than trying to patch all occurrences in the
sources. This should hopefully mean less potential for breakage in
future Discourse releases, too.

talyz commented:
I've added a fix for the `/public` path issue from #142528 that I think is better than the current patch. I recently noticed that backup downloads don't work and it turned out that the root cause was the same issue (a mismatch between how nginx and Discourse refer to the same path in sendfile requests). There may be more places we haven't encountered yet, and even more could crop up in the future, so I've fixed the issue by updating the nginx configuration instead.

ryantm commented:
@talyz that change looks good to me

talyz commented:
I'll merge this on Wednesday (2021-12-08) unless anyone objects. It would be nice to get some more eyes on the v8 change, but it shouldn't affect nodejs (unless it fails the build), since it doesn't touch any of the default outputs.

dpausp reviewed (commented):
Thanks! Code changes look good to me, upgraded my test system from 2.7.9. Tested basic posting functionality, backups and the following plugins: discourse-assign, discourse-canned-replies, discourse-docs, discourse-spoiler-alert, discourse-voting.
